### PR TITLE
NQL parser/Interpreter and database

### PR DESCRIPTION
**Pull request by galyfray**
- created streams :
  - CharStream : exactly what his names implies 
  - TokenStream : take chars and transform them into token through the help of tokenizers
- Tokenizers : class that handle the transformation from char to token, a token is defined by his type and his value
- NQLParser : this class contains everything needed to transform the array of token outputed by the TokenStream into a usable AST ofr the interpreter.
- DatabaseManager : this act as the interpreter of the NQL and will gather the data from the different tablesan then return it
- Created an end point for the server's API : request which require an argument : query with a valid NQL query.
- NQL : 
  - this language is an SQL like. all planned features are not there but they will be added later on.
  - The query is structured around 3 keywords : 
   - SELECT : followed by the column names of what should be selected separated by one to many space. the column names can be prefixed by the the name of the table they belongs to. 
     - TL;DR : `SELECT col1 col2 table.col3`
   - FROM : followed with the name of the table to takes the column ... from .
   - WHERE : (optional) followed by condition. The following element are currently supported :
     - parenthesized expression
     - `AND` and `OR`
     - `=` `!=` `>=` ...
  - Note that the keywords are not case sensitive but the tables and columns name currently are
  - The language is heavily typed it won't accept ton compare string to float and so on 
    - The language won't accept to compare int to float this will be fixed
  - floats use the `.` as the decimal separator 
  - strings should use the simple quote `'` the double won't work 
- CSV parser : 
  - parses CSV real quick 
  - isn't flexible about what's allowed to be or not 
  - only understand the `;` as a separator.
  - only understand the `"` as a quote 
  - doesn't understand what an escapement char is 
- Removed WASM
  - that was a bad idea to try to use it ... 
- Added compression on the server's responses
- switched from HTTP 1.1 to HTTP 2.0 

A lot has been done and there is still a lot to do so I decided to make a PR to stage this and reduce the amount of code to review in one PR.

### CODE REVIEW

File: Server/Ressources/Database/product.csv [hunk 1]
```diff
@@ -0,0 +1,2 @@
+name;price;description;available_quantity;delivery_date;image_url;id;brand;category_id
```
Comment (Yasmarila) [inline Server/Ressources/Database/product.csv]: Maybe we should add an "HT" price and a reference

File: Server/Ressources/Database/category.csv [hunk 2]
```diff
@@ -0,0 +1 @@
+id;name;header_category_id
```
Comment (Yasmarila) [inline Server/Ressources/Database/category.csv:1]: What is header_category_id used for?

File: Server/Ressources/Database/product.csv (hunk 1 above)
Comment (galyfray) [reply Server/Ressources/Database/product.csv]: done

File: Server/Ressources/Database/Columns.csv [hunk 3]
```diff
@@ -0,0 +1,17 @@
+name;type;table_id
+name;string;0
+price;float;0
+description;string;0
+available_quantity;int;0
+delivery_date;date;0
+image_url;string;0
+id;int;0
+brand;string;0
+category_id;int;0
```
Comment (Yasmarila) [inline Server/Ressources/Database/Columns.csv:12]: If we add the HT price & a reference, some changes will be required here

File: Server/Ressources/Database/category.csv (hunk 2 above)
Comment (galyfray) [reply Server/Ressources/Database/category.csv:1]: one category can have multiple sub categories, each sub category only have at most one header category.

Header_category
┗ sub_category

The header's id is here to help making the link between those two elements

File: Server/Ressources/Database/Columns.csv (hunk 3 above)
Comment (galyfray) [reply Server/Ressources/Database/Columns.csv:12]: already done

File: Server/JS/db/NQLParser.js [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+                }
+            }
+        }
+        return value;
+    }
+
+    //=====// Generic helpers //=====//
+
+    /**
+     * Helper function that build an error message an then throw an {@link UnexpectedTokenError}
```
Comment (Yasmarila) [inline Server/JS/db/NQLParser.js:216]: an error message an then  → an error message and then

**Review by Yasmarila — COMMENTED** [review]
Changes in the database

File: Server/JS/db/NQLParser.js (hunk 4 above)
Comment (galyfray) [reply Server/JS/db/NQLParser.js:216]: fixed

**Review by Yasmarila — APPROVED** [review]

**Pull request merged**